dev/com.ibm.ws.jdbc/src/com/ibm/ws/rsadapter/impl/DatabaseHelper.java

```diff
@@ -394,22 +445,24 @@ public boolean isConnectionError(SQLException ex) {
 
         // Maintain a set in order to check for cycles
         Set<Throwable> chain = new HashSet<Throwable>();
-
+        
         boolean stale = false;
         for (Throwable t = ex; t != null && !stale && chain.add(t); t = t.getCause()) {
             SQLException sqlX = t instanceof SQLException ? (SQLException) t : null;
             if (isTraceOn && tc.isDebugEnabled())
                 Tr.debug(this, tc, "checking " + t,
                          sqlX == null ? null : sqlX.getSQLState(),
                          sqlX == null ? null : sqlX.getErrorCode());
+            
             if (sqlX != null)
                 stale |= sqlX instanceof SQLRecoverableException ||
                          sqlX instanceof SQLNonTransientConnectionException ||
                          sqlX instanceof SQLTransientConnectionException && failoverOccurred(sqlX) ||
+                         staleStateAndCodes.contains(sqlX.getSQLState() + "-" + sqlX.getErrorCode()) ||
```

Comment:
This approach is unable to override an improperly identified stale when the JDBC driver raises a SQLNonTransientConnectionException or SQLRecoverableException.
It is also unable to override an improperly identified stale when there was a built-in mapping to just the SQL State, for example 08003 and the user it trying to override so that the combination of SQL State 08003 with error code 230 means not stale. 
That said, we have probably gone too far ahead with the implementation here when we know that ultimately we will also need to consider how other types of identified exceptions fit into the map and may end up altering the approach a bit for that regardless.

Reply:
Makes sense, I've updated the code so user-defined mappings are always checked first, and I also added a testcase called `testRemovedMapping_stateAndCode` which exercises the scenario you mentioned in the second sentence.